doc/compiler/optimizer/ValuePropagation.md

```diff
@@ -0,0 +1,195 @@
+
+Unresolved class constraints are least interesting and express that the class type of the value number has not been loaded by the runtime and is unknown. They simply carry around the string representing the class of the object.
+
+### Null and NonNull Constraints
+
+There are two notions flowed through VP for reference types in regards to nullness. The [null](https://github.com/eclipse/omr/blob/25f479d2b671c2a0a03156cc0a48024a46d2e3c2/compiler/optimizer/VPConstraint.hpp#L746-L760) constraint, and the [non-null](https://github.com/eclipse/omr/blob/25f479d2b671c2a0a03156cc0a48024a46d2e3c2/compiler/optimizer/VPConstraint.hpp#L762-L775) constraint. These constraints are useful for tracking value numbers which are certainly known to be either `null` or certainly not `null`. Knowing such information can eliminate implicit and explicit null checks or open doors for further simplifications.
+
+Note that although these constraint classes exist, we typically query them through the `VPClass` base class constraint. It is usually more convenient and efficient to store all the various information about a reference type in this base class, rather than storing multiple individual constraints for type, nullness, etc.
+
+### Absolute vs. Relative Constraints
+
+Absolute constraints are explicit constraints about a particular value number. For example a type constraint, a range constraint, a null constraint are all examples or absolute constraints.
+
+Relative constraints are constraints that connect one value number to another value number. For example if we were to encounter a guarded comparison between `x` and `y`, say `x < y` then we would create a relative constraint between the value numbers of `x` and `y` to specify that on that path the value number for `x` is less than the value number for `y`.
+
+Relative constraints are useful when neither `x` nor `y` are known to be constant, and we still want to propagate information about the relation of their value numbers. This is useful if for example we encounter another condition which is a successor of the previous check which asked if `x < y + 5`. We want to be able to fold away such expressions even though we don't know any absolute constraints about the value numbers of `x` nor `y`.
+
+## Local vs. Global Value Propagation
+
+There are two modes in which the value propagation optimization can run; local and global. Both modes run using the same constraints and VP handlers. What is different is the scope at which they operate.
+
+Local value propagation processes the program one block at a time. If block extension has ran it processes the program one extended block at a time. Information is not propagated between blocks, so any information we learned is discarded after arriving at the end of the (extended) basic block. This of course makes the local value propagation optimization restrictive in what it can accomplish.
+
```

Comment:
"restrictive" but also "cheaper" from a compile time perspective.

Reply:
edit : you make this point a bit later in the description

Reply:
Fixed in [force-push](https://github.com/eclipse/omr/compare/5c1aa757ca3e116307e60404b9d3b0743e2ce03f..127bdeabc6860a9f211446b05d6210a084512605). I added it just to be extra explicit that the cost is from both memory footprint and compile time.